### ISSUE TITLE
Kart particles effects are displayed multiple times with some settings

### ISSUE THREAD

**Issue by deveee** [supertuxkart/stk-code#1552]
Currently skidding particles are barely visible. I changed rate and life time values in in xml file to higher values. Now I see it twice:

![soccer_field-2014 09 21_10 52 50](https://cloud.githubusercontent.com/assets/5108558/4347843/61325398-416d-11e4-9aaf-38f624dafbdb.png)

**Comment by Flakebi** [comment]
Even if the game runs at high framerates, I see this issue while I'm skidding. It seems like the particles use the direction that the graphical model of the kart is facing, but the real (physical) direction is another because the kart is turned while skidding. So the line of particles that gets created in one frame is far away from the one created in the last frame, because it's created with the wrong direction. That creates the effect that the nitro stream is split into parts.

**Comment by Flakebi** [comment]
Now I'm sure that the divided nitro effect is caused by rotating the visual kart while skidding. The nitro box-emitter is added to the kart node and creates particles every frame, but in the wrong direction (as stated in my previous comment).

I didn't find a way to rotate a box-emitter in irrlicht so I think the only way to fix it is to create another node for the kart that has the same rotation as the physical kart and add the nitro emitters there.

**Comment by hiker** [comment]
Removed invalid, since it appears to be not necessarily related to FPS rate.

**Comment by deveee** [comment]
Kart particles work mostly in the same way as in 0.8.1 version. Simply lifetime values can't be too high. I think that ```lifetime max="50"``` is safe.

I only changed lifetime to:
```
    <lifetime  min="30"
               max="50"          />
```
and it looks fine for me. Though it is still noticeable with high fps values during skidding.

**Comment by hiker** [comment]
The implementation is totally different from 0.8.1. On my hd3000 I see the skid particles too far behind the kart, but they are at the right place when the kart is standing still (with a little patch to always show skid particles). I also see two rows. On the other hand, not really a release blocker, since it works for 'most' cards, so postponed.

**Comment by deveee** [comment]
I mean that I had similar issues in STK 0.8.1 when I was experimenting with different values. I didn't look into code.

Again in STK 0.8.1 when you had very low fps and velocity parameter was different than 0, particles were displayed too far or too high. Personally I workarounded it by setting velocity to 0. But I didn't observe it in current git.

I can reproduce it with intel hd4000 and nvidia 635m. Rather common, but low-fps graphics cards.

I suggest to workaround it by setting lifetime min=30 and max=50. It doesn't look that bad.

**Comment by hiker** [comment]
I had a looong look at the shader, and I think the effect is mostly caused by the parameters used (and the algorithm implemented), and not a real bug.

Sampling the data, I can see that around 80% of all skidding particles are in their first frame, i.e. just emitted - resulting in a kind of static block. The gpu shader does not really emit a certain amount of particles till the right overall number is reached, they are all activated at the same time.

The difference between minimum (15ms) and maximum (25ms) life time is only 10 ms, resulting in most particles having a similar state: either the first frame (just after emitting), or one frame of particle simulation (assuming fps of 60, time step size is around 16 ms - most particles will only act for 1 or 2 frames). The speed values are so low that no particle will move much (0.003 m/ms - with 16 ms time step each particle will move at most 0.05m). At the same time the kart moves around 0.35 m (16 ms * 23 m/s) - so each block of particles at emit time is 0.35 m apart from each other, and the particles will only move 0.05m  - not enough for the the particles in one of the blocks to overlap with the particles from the previous step - resulting in the 'blocked' look we have.

When the kart turns this effect is more visible: when the kart is straight the nitro particles appear in a line of two 'blocks', and it's hard to see if those blocks overlap or not, while the kart is turning, it's much more obvious.

Possible solutions:
- increase the life time for particles (--> more chances for the particles to overlap with particles emitted previously)
- increase speed of particles (--> increase chance for overlap)
- perhaps improve the CPU side to use more than one 'spread' value (spread is just use to rotate the speed vector) - perhaps specify the minimum and maximum speed in each direction:
<velocity x-min="0"  x-max="0.05"  y-min="0" y-max="0"  z-min="-0.01" z-max="-0.05"/>
or so.
- Perhaps additionally improve the shader by using a random number (http://byteblacksmith.com/improvements-to-the-canonical-one-liner-glsl-rand-for-opengl-es-2-0/); or perhaps (to reduce shader overhead) supply one additional pseudo random number from the CPU as uniform to be used (e.g. together with fract(updated_time) to make the particles more random.

Of course, when increasing life time the number of particles need to be reduced ... otherwise karts will look like a fire spying dragon :P

**Comment by hiker** [comment]
Oh, forgot to mention the skidding effect: it appears indeed that skidding (i.e. difference between visual and physical rotation) causes additional artifacts, which we still need to investigate. Code-wise it appears to use the graphical rotation, but I had at least the impression that e.g. nitro comes out at the wrong location when turning (though my graphics driver also has apparently a problem with the shader in that it does not always draw the particles at the right place).

**Comment by deveee** [comment]
>Possible solutions:
>    increase the life time for particles 

Increase life time doesn't help at all. You will see three or more "blocks" of particles instead of two.

>increase speed of particles

This one can help. When you use much higher velocity (eg. ```z="-0.03"```), you can use much lower lifetime to show mostly the same effect. If lifetime is short enough (on my machine ```max="50"``` works fine for 15-20 fps), then you will see all particles in one "block".

We can also use higher value for ```box_z``` to get larger "block" size.

**Comment by Arthur-D** [comment]
The skidding particles seem to not have this issue, but the nitro particles does. See screenshot.
![overworld-2015 06 03_01 54 51](https://cloud.githubusercontent.com/assets/5515261/7949969/7242f2f6-0994-11e5-808c-10a06b81529c.jpg)

**Comment by Flakebi** [comment]
As far as I know there are two or three different issues, which gets a bit confusing.

1. The nitro issue that deveee, Arthur-D and others are talking about: As far as I know (and tried), the problem is that a kart gets an additional rotation when it skids. The nitro emitting node is bound to the kart object so it rotates with the (graphical) kart. That means the particles are emitted in the wrong direction, I didn't find an easy way to change the rotation of a particle emitter without putting it into another object. Rotating the emitter when the kart is skidding should fix this issue.

2. When the alternate pipeline is used, particles look "duplicated" (I only noticed that while skidding), @hiker looked into it for a long time (maybe it has to do something with particle life time), so he's the specialist for it :-)

3. Additionally hiker has an issue (or had?) that the skidding particles are displayed way behind the kart (they should start at the wheels). I know nobody that saw this too so far.

**Comment by hiker** [comment]
I think I understand the nitro particle issue now: it's not really caused by skidding, skidding just appears to make it more obvious.
Nitro particles are released along a line (well, very very thin and longish box), and their coordinates used in the shaders are in world coordinate. So after releasing one 'line' of nitro particles in one frame, the kart moves forward before releasing the next line of particles. The first line of particles remain at the original place since their coordinates are in world coordinates (they move a little bit by the velocity of the particles, but not much).

As long as the kart is driving reasonable straight, those lines partially overlap, and you get the effect of a long, continuous line. Even when the kart is turning a bit, the effect looks fine, since the lines are reasonable close to each other. But when using nitro and skidding, the speed of the kart is faster (i.e. the kart moves more from one frame to the next), and the turn angle of the kart is much bigger. So the lines don't connect or overlap anymore, resulting in the look of several independent nitro lines.

I'll try to modify the shaders so that the particles are simulated in kart space (i.e. relative to the kart), and only at rendering time change the position into world coordinates. This should result in the particles following the kart, and the lines overlapping./connecting even when the kart is skidding. Hopefully the overall effect is still good enough.

**Comment by hiker** [comment]
I've created a branch fix_1552 (77c133a), where I have modified the particle simulation to be done in kart space. This fixes the 'multiple lines' issue, but now the (rather long) nitro line turns with the kart (i.e. it remains a long line, even when the kart turns), which looks completely wrong :)
Not sure in which direction to go: change the nitro parameters to spread them out more, or make them shorter, or try to 'smoothly' move particles taking the kart speed into account (somehow) to simulate the falling back).
Additionally, this appears to fix the skidding nitro effects for me - skidding now looks fine for me ... hooray :)
Feedback and suggestions welcome!

**Comment by hiker** [comment]
Note that I will try to fix the particles in a different way, please ignore branch fix_1552 for now.

**Comment by Elderme** [comment]
Hi,

I had a look at the code, and I think the issue comes from:
 - the computation of particle position when a dead particle is reborn
 - the implementation of Nitro emitter as a line whose length is the distance traveled by the kart

When a particle dies, the new particle has to appear at the same position. As a particle never dies exactly at the instant when we render the scene, but rather between two frames, we have to rebuild the position the emitter had, and deduce the particle position.

When the kart is driving straight, the nitro emitter as a long line works fine as the rebirth position of particles have to be behind the kart. But when the kart turns, or when the kart is skidding, the rebirth position have to be behind and a bit on the left or right.

I've tried to modify the shader code with an additional uniform parameter: the emitter matrix in previous frame. With this information I can compute new particles positions in a way which seem more correct to me. It still needs adjustments but the result looks correct (I've also tried with a 20 fps framerate limit), without multiple lines:
![skidding_nitro](https://cloud.githubusercontent.com/assets/13124234/9173308/9f86922a-3f79-11e5-8f74-0cbedff16220.png)

I will create a branch and a PR.
